### ISSUE TITLE
Suggestion: replace `PaymentWallets` with a property on `PaymentInstrument`

### ISSUE THREAD

**Issue by dlongley** [w3c/webpayments-payment-apps-api#153]
It seems like having two storage containers for payment instruments (`PaymentInstruments` and `PaymentWallets`) is unnecessarily heavy weight. If grouping is desirable, can we instead add an optional `paymentWallet` or `paymentInstrumentGroup` property to the `PaymentInstrument` dictionary?

User agents could use the value of this property to group payment instruments for display in the same way they can now with `PaymentWallet` containers, but it would be much lighter weight and likely more attractive from an implementation perspective.

**Comment by dlongley** [comment]
This approach wouldn't necessarily make icons available for wallets, but they could be grouped under the same "name" which could be displayed:

```js
function addInstruments(registration) {
  const instrumentPromises = [
    registration.paymentManager.instruments.set(
      "dc2de27a-ca5e-4fbd-883e-b6ded6c69d4f",
      {
        name: "Visa ending ****4756",
        enabledMethods: ["basic-card"],
        capabilities: {
          supportedNetworks: ['visa'],
          supportedTypes: ['credit']
        },
        group: "Acme Bank Personal Accounts"
      }),

    registration.paymentManager.instruments.set(
      "c8126178-3bba-4d09-8f00-0771bcfd3b11",
      {
        name: "My Bob Pay Account: john@example.com",
        enabledMethods: ["https://bobpay.com/"],
        group: "Acme Bank Personal Accounts"
      }),

    registration.paymentManager.instruments.set(
      "new-card",
      {
        name: "Add new credit/debit card to ExampleApp",
        enabledMethods: ["basic-card"],
        capabilities: {
          supportedNetworks:
            ['visa','mastercard','amex','discover'],
          supportedTypes: ['credit','debit','prepaid']
        },
        group: "Acme Bank Personal Accounts"
      })
    ];

    return Promise.all(instrumentPromises);
  }
```

Just floating this as a lighter weight approach to getting grouping into the spec if the heavier weight `PaymentWallets` approach seems unlikely to get implemented.

**Comment by adrianhopebailie** [comment]
+1

This would make the spec significantly simpler and resolve the outstanding issue of getting icons for the wallet.

Note also that we have heard from at least one browser that they consider the wallet feature to be adding unnecessary complexity.

**Comment by msporny** [comment]
+1 to removing PaymentWallets in favor of the simpler proposal above. 

If group icon fetching is still an issue, we could always add a ```groupIcon``` property to payment instrument registration. The obvious downside to this is if the developer lists different groupIcons for the same group. A potential solution for this would be that the browser would decide which icon to use (for example, it would use the one that's mentioned the most by the payment instruments in the group and if there is a tie, the lexicographically least icon is used).

**Comment by ianbjacobs** [comment]
-1 to removing the wallet concept. I have spoken to one vendor recently, for example, who indicated they thought that the wallet display option was very compelling as a feature. 

Let's figure out the right balance (after FPWD).

Ian

**Comment by adrianhopebailie** [comment]
> Let's figure out the right balance (after FPWD).

Agree that this can wait until after FPWD

Two points in favor of removing wallets:

1. They can be emulated by using subdomains and therefor payment handlers at multiple origins from the same provider.
2. They can easily be added later either in the fashion @dlongley suggest or as currently specified. We should start simple and add features based on demand

**Comment by zkoch** [comment]
Now that it's after the FPWD, I would propose we come back to this. I also think that PaymentWallets add unnecessary complexity and would like to see them removed (Chrome has no plans to support as is). I'm skeptical that we even need grouping functionality at all, and I would propose we go a step further and remove PaymentWallet without replacement.

While I understand there are some vendors who might see this functionality as compelling, I'm not sure that alone merits the complexity being added.

cc @marcoscaceres

**Comment by marcoscaceres** [comment]
Agree. We should scale back wallets, see what we learn from initial implementation and then see if we actually need them. There might also be simpler ways of grouping payment instruments together - but we should explore that in the future.

**Comment by ianbjacobs** [comment]
The WPWG discussed this today:
 https://www.w3.org/2017/07/20-wpwg-minutes.html#item06

* There is some support for the use cases (e.g., Worldpay, Airbnb) although also some sentiment that they may not be the top priority use cases.
* There is support for addressing these use cases through means other than specific API features (Wallet Dictionary). 
* The first alternative suggest was to use service workers with different scopes and independent Web app manifests. @rsolomakhin said he would be willing to write up an example for the spec about how to do this (once he has time).
* This raised the question about a hard dependency on web app manifest. There was some sentiment on the call that its implementation and spec status would likely remain ahead of payment handler, so this hard dependency was currently perceived as an important risk.
* The second alternative was to use subdomains, which offer web security benefits. However, there was also mention of certificate "cost" with this approach.

In summary, I'd like to propose that we remove the "Wallet" feature but mention in the spec how to fulfill the use cases through other means, including an example from Rouslan.

Ian

**Comment by marcoscaceres** [comment]
> In summary, I'd like to propose that we remove the "Wallet" feature but mention in the spec how to fulfill the use cases through other means, including an example from Rouslan.

**Comment by marcoscaceres** [comment]
(fat fingers!)
> In summary, I'd like to propose that we remove the "Wallet" feature but mention in the spec how to fulfill the use cases through other means, including an example from Rouslan.

I'm worried that the multiple service worker thing... it's already super complicated just dealing with 1. I wouldn't want to see that in the spec - but maybe in the wiki or something.

**Comment by rsolomakhin** [comment]
I've written up [how to use two service workers](https://github.com/rsolomakhin/rsolomakhin.github.io/blob/master/pr/sw.md). It's a bit bare-bones. Does it make sense?

**Comment by ianbjacobs** [comment]
Thank you @rsolomakhin

**Comment by ianbjacobs** [comment]
Discussed at 22 August task force call:
http://www.w3.org/2017/08/22-apps-minutes.html

@rsolomakhin will update examples accordingly.

**Comment by ianbjacobs** [comment]
Hi @rsolomakhin,

I see that you updated the examples:
 https://github.com/rsolomakhin/rsolomakhin.github.io/blob/master/pr/sw.md

Should we migrate these to the developer portal?
 https://github.com/w3c/payment-request-info/wiki/PaymentAppPractice

Ian

**Comment by rsolomakhin** [comment]
Certainly.

**Comment by ianbjacobs** [comment]
With that I am closing this issue as we are not currently planning to support wallets through specific features of this API (but rather through other mechanisms).
